pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,584 @@
+                int_g.copy(source_kernels=(knl,), densities=(density,)),
+                base_kernel)
+    return result
+
+
+def _rewrite_int_g_using_base_kernel(int_g: IntG, base_kernel: ExpressionKernel) \
+        -> ExpressionT:
+    r"""Rewrites an ``IntG`` with only one source kernel to an expression with
+    ``IntG``\ s having the base kernel *base_kernel*.
+    """
+    target_kernel = int_g.target_kernel.replace_base_kernel(base_kernel)
+    dim = target_kernel.dim
+
+    result = 0
+
+    density, = int_g.densities
+    source_kernel, = int_g.source_kernels
+    deriv_relation = get_deriv_relation_kernel(source_kernel.get_base_kernel(),
+        base_kernel, hashable_kernel_arguments=(
+            hashable_kernel_args(int_g.kernel_arguments)))
+
+    const = deriv_relation.const
+    # NOTE: we set a dofdesc here to force the evaluation of this integral
+    # on the source instead of the target when using automatic tagging
+    # see :meth:`pytential.symbolic.mappers.LocationTagger._default_dofdesc`
+    if int_g.source.geometry is None:
+        dd = int_g.source.copy(geometry=DEFAULT_SOURCE)
+    else:
+        dd = int_g.source
+    const *= pytential.sym.integral(dim, dim-1, density, dofdesc=dd)
+
+    if const != 0 and target_kernel != target_kernel.get_base_kernel():
+        # There might be some TargetPointMultipliers hanging around.
+        # FIXME: handle them instead of bailing out
+        return int_g
+
+    if const != 0 and source_kernel != source_kernel.get_base_kernel():
+        # We only handle the case where any source transformation is a derivative
+        # and the constant when applied becomes zero. We bail out if not
+        knl = source_kernel
+        while isinstance(knl, KernelWrapper):
+            if not isinstance(knl,
+                              (AxisSourceDerivative, DirectionalSourceDerivative)):
+                return int_g
+            knl = knl.inner_kernel
+        const = 0
+
+    result += const
+
+    new_kernel_args = filter_kernel_arguments([base_kernel],
+            int_g.kernel_arguments)
+
+    for mi, c in deriv_relation.linear_combination:
+        knl = source_kernel.replace_base_kernel(base_kernel)
+        for d, val in enumerate(mi):
+            for _ in range(val):
+                knl = AxisSourceDerivative(d, knl)
+                c *= -1
+        result += int_g.copy(source_kernels=(knl,), target_kernel=target_kernel,
+                densities=(density * c,), kernel_arguments=new_kernel_args)
+    return result
+
+
+@dataclass
+class DerivRelation:
+    """A class to hold the relationship between a kernel and a base kernel.
+    *linear_combination* is a list of pairs of (mi, coeff).
+    The relation is given by,
+    `kernel = const + sum(deriv(base_kernel, mi) * coeff)`
+    """
+    const: ExpressionT
+    linear_combination: List[Tuple[Tuple[int, ...], ExpressionT]]
+
+
+def get_deriv_relation(kernels: List[ExpressionKernel],
```

Comment:
Write down the math that is going on under the hood, at least roughly.

Reply:
Done in a407f00